### PR TITLE
IP APIs sanity

### PR DESCRIPTION
**Pull request by tbursztyka**

### CODE REVIEW

**Comment by tbursztyka** [comment]
@ozhuraki These patches are doing what you attempted in #8561 
As you did, _raw should disappear but also the context based functions should go in net_context.c

(Well I have to fix the mld test, will do that)

File: subsys/net/ip/tcp.c [hunk 1]
```diff
@@ -361,12 +361,14 @@ static int finalize_segment(struct net_context *context, struct net_pkt *pkt)
...
 	if (net_pkt_family(pkt) == AF_INET) {
-		return net_ipv4_finalize(context, pkt);
+		return net_ipv4_finalize(pkt,
+					 net_context_get_ip_proto(context));
 	} else
 #endif
 #if defined(CONFIG_NET_IPV6)
 	if (net_pkt_family(pkt) == AF_INET6) {
-		return net_ipv6_finalize(context, pkt);
+		return net_ipv6_finalize(pkt,
```
Comment (tbursztyka) [inline subsys/net/ip/tcp.c]: Hum, I mixed up ipv6 changes in ipv4 patch...

File: subsys/net/ip/net_context.c [hunk 2]
```diff
@@ -623,6 +623,61 @@ int net_context_listen(struct net_context *context, int backlog)
...
+	NET_ASSERT(ipv4);
+	NET_ASSERT(((struct sockaddr_in_ptr *)&context->local)->sin_addr);
+
+	if (!src) {
+		src = ((struct sockaddr_in_ptr *)&context->local)->sin_addr;
+	}
+
+	if (net_is_ipv4_addr_unspecified(src)
+	    || net_is_ipv4_addr_mcast(src)) {
+		src = &ipv4->unicast[0].address.in_addr;
```
Comment (jukkar) [inline subsys/net/ip/net_context.c]: We should change this to use the net_if_ipv4_select_src_addr() that was introduced recently.

**Review by jukkar — CHANGES_REQUESTED** [review]

**Review by pfalcon — APPROVED** [review]
This is definitely a step in the right direction.

**Review by jukkar — APPROVED** [review]

**Pull request merged**